File: source/python/ring_buffer.py
class RingBuffer:
    def __init__(self, capacity):
        self.capacity = capacity
        self.data = [None] * capacity
        self.head = 0
        self.tail = 0
        self.count = 0

    def write(self, val):
        if self.count == self.capacity:
            print(f"ring full, cannot write {val}")
            return False
        self.data[self.tail] = val
        self.tail = (self.tail + 1) % self.capacity
        self.count += 1
        return True

    def read(self):
        if self.count == 0:
            print("ring empty")
            return None
        val = self.data[self.head]
        self.head = (self.head + 1) % self.capacity
        self.count -= 1
        return val

rb = RingBuffer(4)
rb.write(1)
rb.write(2)
rb.write(3)
rb.write(4)

results = []
results.append(rb.read())
results.append(rb.read())

rb.write(5)
rb.write(6)

results.append(rb.read())
results.append(rb.read())
results.append(rb.read())
results.append(rb.read())

print(" ".join(str(x) for x in results))

File: source/rust/ring_buffer.rs
struct RingBuffer {
    data: Vec<i32>,
    head: usize,
    tail: usize,
    count: usize,
    capacity: usize,
}

impl RingBuffer {
    fn new(capacity: usize) -> Self {
        RingBuffer { data: vec![0; capacity], head: 0, tail: 0, count: 0, capacity }
    }

    fn write(&mut self, val: i32) -> bool {
        if self.count == self.capacity {
            println!("ring full, cannot write {}", val);
            return false;
        }
        self.data[self.tail] = val;
        self.tail = (self.tail + 1) % self.capacity;
        self.count += 1;
        true
    }

    fn read(&mut self) -> Option<i32> {
        if self.count == 0 {
            println!("ring empty");
            return None;
        }
        let val = self.data[self.head];
        self.head = (self.head + 1) % self.capacity;
        self.count -= 1;
        Some(val)
    }
}

fn main() {
    let mut rb = RingBuffer::new(4);
    rb.write(1); rb.write(2); rb.write(3); rb.write(4);
    let mut results = Vec::new();
    results.push(rb.read().unwrap());
    results.push(rb.read().unwrap());
    rb.write(5); rb.write(6);
    results.push(rb.read().unwrap());
    results.push(rb.read().unwrap());
    results.push(rb.read().unwrap());
    results.push(rb.read().unwrap());
    let s: Vec<String> = results.iter().map(|x| x.to_string()).collect();
    println!("{}", s.join(" "));
}
